asdf/compression.py

```diff
@@ -30,118 +34,232 @@ def validate(compression):
+                        self._size = 0
+                else:
+                    # We have at least one full block
+                    _out = self._api.decompress(memoryview(block[:self._size]), return_bytearray=True)
+                    out[bytesout:bytesout+len(_out)] = _out
+                    bytesout += len(_out)
+                    block = block[self._size:]
+                    self._size = 0
+
+        return bytesout
+
+    
+class ZlibCompressor:
+    def __init__(self, compression_kwargs=None, decompression_kwargs=None):
+        if compression_kwargs is None:
+            compression_kwargs = {}
+        if decompression_kwargs is None:
+            decompression_kwargs = {}
+            
+        self.compression_kwargs = compression_kwargs.copy()
+        self.decompression_kwargs = decompression_kwargs.copy()
+        
+    def compress(self, data, out=None):
+        comp = zlib.compress(data, **self.compression_kwargs)
+        if out is not None:
+            out[:len(comp)] = comp
+            return len(comp)
+        return comp
+    
+    def decompress(self, blocks, out):
+        decompressor = zlib.decompressobj(**self.decompression_kwargs)
+        
+        i = 0
+        for block in blocks:
+            decomp = decompressor.decompress(block)
+            out[i:i+len(decomp)] = decomp
+            i += len(decomp)
+        return i
+
+
+class Bzp2Compressor:
+    def __init__(self, compression_kwargs=None, decompression_kwargs=None):
+        if compression_kwargs is None:
+            compression_kwargs = {}
+        if decompression_kwargs is None:
+            decompression_kwargs = {}
+            
+        self.compression_kwargs = compression_kwargs.copy()
+        self.decompression_kwargs = decompression_kwargs.copy()
+        
+    def compress(self, data, out=None):
+        comp = bz2.compress(data, **self.compression_kwargs)
+        if out is not None:
+            out[:len(comp)] = comp
+            return len(comp)
+        return comp
+    
+    def decompress(self, blocks, out):
+        decompressor = bz2.BZ2Decompressor(**self.decompression_kwargs)
+
+        i = 0
+        for block in blocks:
+            decomp = decompressor.decompress(block)
+            out[i:i+len(decomp)] = decomp
+            i += len(decomp)
+        return i
+
+        
+def _get_compressor_from_extensions(compression, return_extension=False):
+    '''
+    Look at the loaded ASDF extensions and return the first one (if any)
+    that can handle this type of compression.
+    `return_extension` can be used to return corresponding extension for bookkeeping purposes.
+    Returns None if no match found.
+    '''
+    # TODO: in ASDF 3, this will be done by the ExtensionManager
+    extensions = get_config().extensions
+
+    for ext in extensions:
+        for comp in ext.compressors:
+            # TODO: slightly unfortunate to have to construct the object to get the labels
```

Comment:
I mentioned this up above, but we might want to return the label from a class method